src/components/utils/LayerManager.ts

```diff
@@ -17,27 +18,163 @@ export type ContentElement =
     | (VirtualElement & {contains?: (other: Node | null) => boolean});
 
 export interface LayerConfig extends LayerExtendableProps {
+    idle?: boolean;
+    idleTimeout?: number;
+    idlePriority?: number;
     contentRefs?: Array<React.RefObject<ContentElement> | undefined>;
+    onRemove?: () => void;
 }
 
-class LayerManager {
+const createPromise = (): {promise: Promise<void>; resolve: () => void; reject: () => void} => {
+    let resolve!: () => void;
+    let reject!: () => void;
+
+    const promise = new Promise<void>((_resolve, _reject) => {
+        resolve = _resolve;
+        reject = _reject;
+    });
+
+    return {promise, resolve, reject};
+};
+
+type LayerCandidate = {
+    config: LayerConfig;
+    promise: Promise<void>;
+    resolve: () => void;
+    reject: () => void;
+};
+
+const defaultIdleTimeout = 1000;
+
+const compareCandidatesByPriority = (candA: LayerCandidate, candB: LayerCandidate) => {
+    const priorityA = candA.config.idlePriority;
+    const priorityB = candB.config.idlePriority;
+    const isAInt = Number.isInteger(priorityA);
+    const isBInt = Number.isInteger(priorityB);
+
+    if ((priorityA === undefined && priorityB === undefined) || (!isAInt && !isBInt)) {
+        return 0;
+    }
+
+    if (priorityA === undefined || (isBInt && !isAInt)) {
+        return 1;
+    }
+
+    if (priorityB === undefined || (isAInt && !isBInt)) {
+        return -1;
+    }
+
+    return Math.sign(priorityB - priorityA);
+};
+
+export class LayerManager {
     private stack: LayerConfig[] = [];
+    private preStack: LayerCandidate[] = [];
     private mouseDownLayerTarget?: {layer: LayerConfig; target: HTMLElement};
+    private scheduledAdd: ReturnType<typeof setTimeout> | undefined;
+    private scheduledCandidate: LayerCandidate | undefined;
 
-    add(config: LayerConfig) {
-        this.stack.push(config);
+    private checkPreStack = _debounce(() => {
+        if (this.stack.length !== 0) {
+            return;
+        }
 
-        if (this.stack.length === 1) {
-            this.addListeners();
+        const sortedByPriorityDesc = this.preStack.slice();
+        sortedByPriorityDesc.sort(compareCandidatesByPriority);
+        const candidate = sortedByPriorityDesc.shift();
+
+        if (!candidate || candidate === this.scheduledCandidate) {
+            return;
+        }
+
+        this.unscheduleCandidate();
+
+        const scheduledAdd = (scheduledCandidate: LayerCandidate) => {
+            const index = this.preStack.indexOf(scheduledCandidate);
+            this.preStack.splice(index, 1);
+            this.addToStack(scheduledCandidate.config);
+            scheduledCandidate.resolve();
+            this.unscheduleCandidate();
+        };
+
+        const timeout = Number.isInteger(candidate.config.idleTimeout)
+            ? candidate.config.idleTimeout
+            : defaultIdleTimeout;
+
+        this.scheduledAdd = setTimeout(scheduledAdd.bind(this, candidate), timeout);
```

Comment:
```suggestion
        this.scheduledAdd = setTimeout(scheduledAdd.bind(null, candidate), timeout);
```

Reply:
Wow, I didn't know it works like that – this `bind(...)` seems to completely ignore `null` because scheduledAdd is an arrow function (already bound).